examples/timeseries/timeseries_transformer_classification.py

```diff
@@ -0,0 +1,189 @@
+        x = self.attention(x, **kwargs)
+        x = self.attention_dropout(x, training=training, **kwargs)
+        res = x + inputs
+
+        x = self.ff_norm(res, **kwargs)
+        x = self.ff_conv1(x, **kwargs)
+        x = self.ff_dropout(x, training=training, **kwargs)
+        x = self.ff_conv2(x, **kwargs)
+        x = self.ff_dropout(x, training=training, **kwargs)
+        return x + res
+
+    def compute_output_shape(self, input_shape):
+        return input_shape
+
+"""
+Finally, we can wrap every layer up and add the ability to stack several of them together.
+"""
+
+class TransformerEncoder(keras.Model):
+    def __init__(self, name='TransformerEncoder', num_heads=1, head_size=64, ff_dim=None, num_layers=1, dropout=0, **kwargs):
+        super().__init__(name=name, **kwargs)
+        if ff_dim is None: # sensible default
+            ff_dim = head_size
+        self.dropout = dropout
+        self.attention_layers = [SelfAttentionBlock(
+            num_heads=num_heads, head_size=head_size, ff_dim=ff_dim, dropout=dropout) for _ in range(num_layers)]
+
+    def call(self, inputs, training, **kwargs):
+        x = inputs # (batch size, sequence length, features)
+        for attention_layer in self.attention_layers:
+            x = attention_layer(x, training, **kwargs)
+
+        return x # (batch size, sequence length, features)
+
+"""
+The main part of our model is now complete and we can now proceed to add the final Multi-Layer Perceptron classification "head".
+Apart from a stack of `Dense` layers, we need to reduce the output tensor of the `TransformerEncoder` part of our model,
+down to a vector of features for each data point in the current batch.
+A common way to achieve this is to use some sort of pooling layer.
+For this example, a `GlobalAveragePooling1D` layer is sufficient.
+Feel free to experiment with different pooling layers if you wish, encorporating domain expertise.
+"""
+
+def build_model(input_shape, n_classes, head_size, num_heads=4, num_layers=2, dropout=0.2, dropout_mlps=0.5, mlp=[],ff_dim=None, activation='relu', optimizer='adamW'):
+    model_input = keras.Input(shape=input_shape)
+    x = model_input
+    x = TransformerEncoder(num_heads=num_heads, head_size=head_size, ff_dim=None, num_layers=num_layers, dropout=dropout)(x)
+
+    # if you use the default format 'channels_last', then the output dimensions
+    # will be very small, of shape (seq len, 1)
+    head = keras.layers.GlobalAveragePooling1D(data_format='channels_first')(x)
+
+    for dim in mlp:
+        head = keras.layers.Dense(dim, activation=activation)(head)
+        head = keras.layers.Dropout(dropout_mlps)(head)
+
+    model_output = keras.layers.Dense(n_classes, activation='softmax')(head)
+
+    model = keras.Model(model_input, model_output)
+
+    model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizer, metrics=['sparse_categorical_accuracy'])
+    model.summary()
+    return model
```

Comment:
The entire modeling part of this example can be rewritten as:

```python
def transformer_encoder(inputs, head_size, num_heads, ff_dim, dropout=0):
    x = layers.LayerNormalization(epsilon=1e-6)(inputs)
    x = layers.MultiHeadAttention(
        key_dim=head_size, num_heads=num_heads, dropout=dropout)(x, x)
    x = layers.Dropout(dropout)(x)
    res = x + inputs
    x = layers.LayerNormalization(epsilon=1e-6)(res)
    x = layers.Conv1D(filters=ff_dim, kernel_size=1, activation='relu')(x)
    x = layers.Dropout(dropout)(x)
    x = layers.Conv1D(filters=inputs.shape[-1], kernel_size=1)(x)
    return x + res

def build_model(input_shape, ...):
    inputs = keras.Input(shape=input_shape)
    x = inputs
    for _ in range(num_blocks):
      x = transformer_encoder(x, ...)
    x = keras.layers.GlobalAveragePooling1D(data_format='channels_first')(x)
    for dim in mlp:
        x = keras.layers.Dense(dim, activation=activation)(x)
        x = keras.layers.Dropout(dropout_mlps)(x)
    outputs = keras.layers.Dense(n_classes, activation='softmax')(x)
    return keras.Model(inputs, outputs)
```

Which is a lot more concise. You don't need 3 separate classes and a function. In fact you don't need any classes at all.

Remember that Keras is supposed to be simple. If you find yourself writing a lot of code, there's almost certainly a better, simpler way. Always go with the simplest thing you can do.

Reply:
Thank you for the code review. I made everything simpler and added 2 small paragraphs as introduction and conclusion.